fair/forward.py

```diff
@@ -1,326 +1,309 @@
-  # sublime snippet for variable description is 'vardesc'
-  # fig: (matplotlib axes object)
-  #   the matplotlib axes with all our plotted variables
-
-  # # ------------ SIDE EFFECTS ------------ # #
-  # document side effects here
-
-  # # ------------ EXCEPTIONS ------------ # #
-  # sublime snippet for exception description is 'excdesc'
-
-  # # ------------ RESTRICTIONS ------------ # #
-  Document any restrictions on when the function can be called
-
-  """
-
-  # One line break before anything else
-
-  # # # ------------ IMPORT REQUIRED MODULES ------------ # # #
-  # # ------------ STANDARD LIBRARY ------------ # #
-  from math import ceil
-
-  # # ------------ THIRD PARTY ------------ # #
-  import numpy as np
-
-  from matplotlib import pyplot as plt
-  plt.style.use('seaborn-darkgrid')
-  plt.rcParams['figure.figsize'] = 16, 9
-  plt.rcParams['lines.linewidth'] = 1.5
-
-  font = {'weight' : 'normal',
-          'size'   : 12}
-
-  plt.rc('font', **font)
-
-  # # ------------ LOCAL APPLICATION/LIBRARY SPECIFIC ------------ # #
-
-  # # # ------------ CODE ------------ # # #
-  # # ------------ SORT OUT INPUT VARIABLES ------------ # #
-  pts = {'emms':emms,
-         'forc':forc,
-         'conc':conc,
-         'temp':temp}
-  
-  integ_len = 0
-
-  for j,var in enumerate(pts):
-    if type(pts[var]) == list:
-      pts[var] = np.array(pts[var])
-      integ_len = len(pts[var])
-    elif type(pts[var]) == np.ndarray:
-      integ_len = len(pts[var])
-
-  if integ_len == 0:
-    for name in pts:
-      print "{0}: {1}".format(name,type(pts[name]))
-    raise ValueError("Error: I can't work out which one of your input variables is a timeseries")
-
-  for j,var in enumerate(pts):
-    if type(pts[var]) == np.ndarray and len(pts[var]) == integ_len:
-      pass
-    elif type(pts[var]) == np.ndarray and len(pts[var]) != integ_len:
-      for name in pts:
-        print "{0}: {1}\nlength: {2}".format(name,
-                                             type(pts[name]),
-                                             len(pts[name]))
-      raise ValueError("Error: Your timeseries are not all the same length, I don't know what to do")
+    if np.isscalar(other_rf):
+      F[0,0] = (F2x/np.log(2.)) * np.log(
+        (C[0,0] + C_pi[0]) / C_pi[0]) + other_rf
     else:
-      if type(pts[var]) in [float,int]:
-        pts[var] = np.full(integ_len,pts[var])
+      F[0,0] = (F2x/np.log(2.)) * np.log(
+        (C[0,0] + C_pi[0]) / C_pi[0]) + other_rf[0]
+
+  if restart_in == False:
+    # Update the thermal response boxes
+    T_j[0,:] = (q/d)*(np.sum(F[0,:]))
+
+  # Sum the thermal response boxes to get the total temperature anomaly
+  T[0]=np.sum(T_j[0,:],axis=-1)
+
+  for t in range(1,nt):
+    # Calculate the parametrised iIRF and check if it is over the maximum 
+    # allowed value
+    iirf[t] = rc * C_acc[t-1]  + rt*T[t-1]  + r0
+    if iirf[t] >= iirf_max:
+      iirf[t] = iirf_max
+      
+    # Linearly interpolate a solution for alpha
```

Comment:
I'm guessing the old solver was too slow?

Reply:
This one was Richard's improvement I think. I used `scipy.newton` (fast but didn't always find the root) then `scipy.brentq` if that failed (slower but is guaranteed to find the root if one exists). I believe `scipy.root` does this kind of donkey work automatically.

Reply:
Clever, I think this is still the slowest part of our runs? If that's the case then perhaps there's an even faster way shouuld performance eveer become an issue